app/models/planning_application.rb

```diff
@@ -103,11 +126,52 @@ class PlanningApplication < ApplicationRecord
                            returned], to: :withdrawn, after: proc { |comment|
                                                                update!(cancellation_comment: comment)
                                                              }
+      after do
+        audit("withdrawn", cancellation_comment)
+      end
     end
 
     after_all_transitions :timestamp_status_change # FIXME: https://github.com/aasm/aasm#timestamps
   end
 
+  def audit_created
+    audit("created", nil, Current.user&.name || Current.api_user&.name)
+  end
+
+  # we already have an auditable module but the arguments need to change. Worth calling super or not?
+  def audit(activity_type, audit_comment = nil, activity_information = nil, _api_user = nil)
+    Audit.create!(
+      planning_application_id: id,
+      user: Current.user,
+      audit_comment: audit_comment,
+      activity_information: activity_information,
+      activity_type: activity_type,
+      api_user: Current.api_user
+    )
+  end
+
+  def audit_update_actions
+    saved_changes.keys.map do |attribute_name|
+      case attribute_name
+      when "constraints"
+        prev_arr, new_arr = saved_changes[:constraints]
+
+        attr_removed = prev_arr - new_arr
+        attr_added = new_arr - prev_arr
+
+        attr_added.each { |attr| audit("constraint_added", attr) }
+        attr_removed.each { |attr| audit("constraint_removed", attr) }
+      when "boundary_geojson"
+        audit("red_line_updated", "Red line drawing updated")
+      else
+        audit("updated",
+              "Changed from: #{saved_change_to_attribute(attribute_name).first}
+                \r\n Changed to: #{saved_change_to_attribute(attribute_name).second}",
+              attribute_name.humanize)
+      end
+    end
+  end
```

Comment:
This seems to be doing quite a lot of stuff.
Could we make sure the test coverage is high for different scenarios here (if not already)
Will this run for every update action? So even like address / contact info change on the planning application would hit the else block? Is this what we are supposed to be doing